torchdata/datapipes/iter/util/unzipper.py

```diff
@@ -65,11 +66,46 @@ def __new__(
             )
 
         # The implementation basically uses Forker but only yields a specific element within the sequence
-        container = _UnZipperIterDataPipe(source_datapipe, sequence_length, buffer_size)  # type: ignore[arg-type]
-        return [_ChildDataPipe(container, i) for i in instance_ids]
+        container = _UnZipperIterDataPipe(source_datapipe, instance_ids, buffer_size)  # type: ignore[arg-type]
+        return [_ChildDataPipe(container, i) for i in range(len(instance_ids))]
 
 
 class _UnZipperIterDataPipe(_ForkerIterDataPipe):
+    def __init__(self, datapipe: IterDataPipe, instance_ids: List[int], buffer_size: int = 1000):
+        super().__init__(datapipe, len(instance_ids), buffer_size)  # type: ignore[arg-type]
+        self.instance_ids = instance_ids
+
     def get_next_element_by_instance(self, instance_id: int):
+        r"""
+        Note:
+            Each element returned from the source datapipe is required to be a sequnce that can
+            be subscribed with a column index
+        """
         for return_val in super().get_next_element_by_instance(instance_id):
-            yield return_val[instance_id]
+            yield return_val[self.instance_ids[instance_id]]
+
+    def __getstate__(self):
+        if IterDataPipe.getstate_hook is not None:
+            return IterDataPipe.getstate_hook(self)
+
+        state = (
+            self.main_datapipe,
+            self.num_instances,
+            self.buffer_size,
+            self.instance_ids,
+        )
+        return state
```

Comment:
I think it would be better to call `super().__getstate__`  and add `self.instance_ids` to it, because now any changes to `Forker`'s `__getstate__` will need to repeat here as well.

Reply:
I can also make this change in a separate PR if this is landing already.

Reply:
Makes sense. Will change it in this PR.